sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala

```diff
@@ -157,31 +159,61 @@ abstract class AggregationIterator(
+      val notExistsFilter = !filterExpressions.exists(_ != None)
+      var isFinalOrMerge = false
+      val mergeExpressions = functions.zipWithIndex.collect {
+        case (ae: DeclarativeAggregate, i) =>
+          expressions(i).mode match {
             case Partial | Complete => ae.updateExpressions
-            case PartialMerge | Final => ae.mergeExpressions
+            case PartialMerge | Final =>
+              isFinalOrMerge = true
+              ae.mergeExpressions
           }
         case (agg: AggregateFunction, _) => Seq.fill(agg.aggBufferAttributes.length)(NoOp)
       }
       val updateFunctions = functions.zipWithIndex.collect {
         case (ae: ImperativeAggregate, i) =>
           expressions(i).mode match {
-            case Partial | Complete =>
+            case Partial | Complete if filterExpressions(i).isDefined =>
+              (buffer: InternalRow, row: InternalRow) =>
+                if (predicates(i).eval(row)) { ae.update(buffer, row) }
+            case Partial | Complete if filterExpressions(i).isEmpty =>
               (buffer: InternalRow, row: InternalRow) => ae.update(buffer, row)
             case PartialMerge | Final =>
               (buffer: InternalRow, row: InternalRow) => ae.merge(buffer, row)
           }
       }.toArray
       // This projection is used to merge buffer values for all expression-based aggregates.
       val aggregationBufferSchema = functions.flatMap(_.aggBufferAttributes)
-      val updateProjection =
-        newMutableProjection(mergeExpressions, aggregationBufferSchema ++ inputAttributes)
+      val updateProjection = newMutableProjection(
+        mergeExpressions.flatMap(_.seq), aggregationBufferSchema ++ inputAttributes)
 
       (currentBuffer: InternalRow, row: InternalRow) => {
         // Process all expression-based aggregate functions.
-        updateProjection.target(currentBuffer)(joinedRow(currentBuffer, row))
+        if (notExistsFilter || isFinalOrMerge) {
+          updateProjection.target(currentBuffer)(joinedRow(currentBuffer, row))
+        } else {
+          val dynamicMergeExpressions = new ArrayBuffer[Expression]
+          for (i <- 0 until expressions.length) {
+            if ((expressions(i).mode == Partial || expressions(i).mode == Complete)) {
+              if (filterExpressions(i).isDefined) {
+                if (predicates(i).eval(row)) {
+                  dynamicMergeExpressions ++= mergeExpressions(i)
+                } else {
+                  dynamicMergeExpressions ++= Seq(NoOp)
+                }
+              } else {
+                dynamicMergeExpressions ++= mergeExpressions(i)
+              }
+            }
+          }
+          if (!dynamicMergeExpressions.isEmpty) {
+            val dynamicUpdateProjection = newMutableProjection(
+              dynamicMergeExpressions, aggregationBufferSchema ++ inputAttributes)
+            dynamicUpdateProjection.target(currentBuffer)(joinedRow(currentBuffer, row))
+          }
```

Comment:
Could you explain about what you do in this code block? Also, plz leave some comments about that.

Reply:
```
 // In the list of aggregate expressions, if a filter predicate is specified for at least one
        // aggregate expression and aggregate expressions are in partial or complete mode,
        // then the filter will be used.
        // Suppose there is a list of aggregate expressions, such as exprA with filterA, exprB, exprC with filterC,
        // then the specific implementation process is as follows:
        // 1. Accept data row.
        // 2. Execute multiple aggregate expressions in sequence.
        // 2-1. Filter the data row using filter predicate filterA. If the filter predicate filterA is met,
        //      then calculate using aggregate expression exprA.
        // 2-2. Calculate using aggregate expression exprB.
        // 2-3. Filter the data row using filter predicate filterC. If the filter predicate filterC is met,
        //      then calculate using aggregate expression exprC.
```